File: Solution Dijkstra Search.js

/**
 * @param {number} numberOfCities
 * @param {number[][]} flights
 * @param {number} start
 * @param {number} goal
 * @param {number} maxNumberOfStops
 * @return {number}
 */
var findCheapestPrice = function (numberOfCities, flights, start, goal, maxNumberOfStops) {
    this.graph = new Map(); //Map<number, City[]>
    this.NO_SUCH_ROUTE = -1;
    initializeGraph(flights);
    return dijkstraSearchForCheapestPriceWithMaxStops(start, goal, numberOfCities, maxNumberOfStops);
};

/**
 * @param {number} ID
 * @param {number} priceOfFlight
 */
function City(ID, priceOfFlight) {
    this.ID = ID;
    this.priceOfFlight = priceOfFlight;
}

/**
 * @param {number} cityID
 * @param {number} priceFromStart
 * @param {number} stopsFromStart 
 */
function Flight(cityID, priceFromStart, stopsFromStart) {
    this.cityID = cityID;
    this.priceFromStart = priceFromStart;
    this.stopsFromStart = stopsFromStart;
}

/**
 * @param {number} start
 * @param {number} goal
 * @param {number} numberOfCities
 * @param {number} maxNumberOfStops
 * @return {number}
 */
function dijkstraSearchForCheapestPriceWithMaxStops(start, goal, numberOfCities, maxNumberOfStops) {
    //const {MinPriorityQueue} = require('@datastructures-js/priority-queue');
    //MinPriorityQueue<Flight>
    const minHeap = new MinPriorityQueue({compare: (x, y) => x.priceFromStart - y.priceFromStart});
    minHeap.enqueue(new Flight(start, 0, 0));

    const stopsFromStart = new Array(numberOfCities).fill(Number.MAX_SAFE_INTEGER);

    while (!minHeap.isEmpty()) {

        const flight = minHeap.dequeue();
        if (flight.cityID === goal) {
            return flight.priceFromStart;
        }
        if (flight.stopsFromStart > stopsFromStart[flight.cityID] || flight.stopsFromStart > maxNumberOfStops) {
            continue;
        }
        stopsFromStart[flight.cityID] = flight.stopsFromStart;
        if (!this.graph.has(flight.cityID)) {
            continue;
        }
        const nextCities = this.graph.get(flight.cityID);
        for (let nextCity of nextCities) {
            minHeap.enqueue(new Flight(nextCity.ID, flight.priceFromStart + nextCity.priceOfFlight, 1 + flight.stopsFromStart));
        }
    }
    return this.NO_SUCH_ROUTE;
}

/**
 * @param {number[][]} flights
 * @return {void}
 */
function initializeGraph(flights) {
    for (let [from, to, price] of flights) {
        if (!this.graph.has(from)) {
            this.graph.set(from, []);
        }
        this.graph.get(from).push(new City(to, price));
    }
}

File: Solution Dijkstra Search.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class Solution {

    private record City(int ID, int priceOfFlight) {}
    private record Flight(int cityID, int priceFromStart, int stopsFromStart) {}
    private Map<Integer, List<City>> graph;
    private static final int NO_SUCH_ROUTE = -1;

    public int findCheapestPrice(int numberOfCities, int[][] flights, int start, int goal, int maxNumberOfStops) {
        initializeGraph(flights);
        return dijkstraSearchForCheapestPriceWithMaxStops(start, goal, numberOfCities, maxNumberOfStops);
    }

    private int dijkstraSearchForCheapestPriceWithMaxStops(int start, int goal, int numberOfCities, int maxNumberOfStops) {
        PriorityQueue<Flight> minHeap = new PriorityQueue<>((x, y) -> x.priceFromStart - y.priceFromStart);
        minHeap.add(new Flight(start, 0, 0));

        int[] stopsFromStart = new int[numberOfCities];
        Arrays.fill(stopsFromStart, Integer.MAX_VALUE);

        while (!minHeap.isEmpty()) {
            Flight flight = minHeap.poll();
            if (flight.cityID == goal) {
                return flight.priceFromStart;
            }
            if (flight.stopsFromStart > stopsFromStart[flight.cityID] || flight.stopsFromStart > maxNumberOfStops) {
                continue;
            }
            stopsFromStart[flight.cityID] = flight.stopsFromStart;
            if (!graph.containsKey(flight.cityID)) {
                continue;
            }
            List<City> nextCities = graph.get(flight.cityID);
            for (City nextCity : nextCities) {
                minHeap.add(new Flight(nextCity.ID, flight.priceFromStart + nextCity.priceOfFlight, 1 + flight.stopsFromStart));
            }
        }
        return NO_SUCH_ROUTE;
    }

    private void initializeGraph(int[][] flights) {
        graph = new HashMap<>();
        for (int[] flight : flights) {
            int from = flight[0];
            int to = flight[1];
            int price = flight[2];
            graph.putIfAbsent(from, new ArrayList<>());
            graph.get(from).add(new City(to, price));
        }
    }
}
